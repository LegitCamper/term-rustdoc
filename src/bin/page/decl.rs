pub struct Declaration {}

impl std::fmt::Debug for Declaration {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Declaration")
    }
}

impl Declaration {}
