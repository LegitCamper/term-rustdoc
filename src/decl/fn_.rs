use crate::{tree::IDMap, type_name::short};
use itertools::Itertools;
use rustdoc_types::{Abi, FnDecl, Function, Header, ItemEnum, Type, Visibility};
use std::fmt::Write;
use std::format_args as f;

fn vis(v: &Visibility, buf: &mut String) {
    match v {
        Visibility::Public => buf.push_str("pub "),
        Visibility::Default => (),
        Visibility::Crate => buf.push_str("pub(crate) "),
        Visibility::Restricted { path, .. } => write!(buf, "pub({path}) ").unwrap(),
    };
}

fn header(h: &Header, buf: &mut String) {
    let Header {
        const_,
        unsafe_,
        async_,
        abi,
    } = h;
    if *const_ {
        buf.push_str("const ");
    }
    if *unsafe_ {
        buf.push_str("unsafe ");
    }
    if *async_ {
        buf.push_str("async ");
    }
    match abi {
        Abi::Rust => (),
        Abi::C { .. } => buf.push_str("extern \"C\""),
        Abi::Cdecl { .. } => buf.push_str("extern \"cdecl\""),
        Abi::Stdcall { .. } => buf.push_str("extern \"stdcall\""),
        Abi::Fastcall { .. } => buf.push_str("extern \"fastcall\""),
        Abi::Aapcs { .. } => buf.push_str("extern \"aapcs\""),
        Abi::Win64 { .. } => buf.push_str("extern \"win64\""),
        Abi::SysV64 { .. } => buf.push_str("extern \"sysv64\""),
        Abi::System { .. } => buf.push_str("extern \"system\""),
        Abi::Other(s) => buf.push_str(s),
    };
}

fn fndecl(f: &FnDecl, buf: &mut String) {
    buf.push('(');
    let iter = f.inputs.iter().enumerate();
    let args = iter.format_with(", ", |(idx, (name, ty)), f| {
        if idx == 0 && name == "self" {
            match ty {
                Type::BorrowedRef {
                    lifetime,
                    mutable,
                    type_,
                } => {
                    let mut ty = short(type_).unwrap_or_default();
                    if ty == "Self" {
                        ty.clear();
                        ty.push_str("self");
                    }
                    return match (lifetime, mutable) {
                        (None, false) => f(&f!("&{ty}")),
                        (None, true) => f(&f!("&mut {ty}")),
                        (Some(life), false) => f(&f!("&{life} {ty}")),
                        (Some(life), true) => f(&f!("&{life} mut {ty}")),
                    };
                }
                Type::Generic(s) if s == "Self" => return f(&"self"),
                _ => (),
            }
        }
        let ty = short(ty).unwrap_or_default();
        f(&f!("{name}: {ty}"))
    });
    write!(buf, "{args}").unwrap();
    buf.push(')');
    if let Some(ty) = &f.output {
        buf.push_str(" -> ");
        if let Some(output) = short(ty) {
            buf.push_str(&output);
        }
    }
}

fn parse_fn(v: &Visibility, fname: &str, fn_item: &Function) -> String {
    let mut buf = String::with_capacity(128);
    vis(v, &mut buf);
    buf.push_str(fname);
    header(&fn_item.header, &mut buf);
    fndecl(&fn_item.decl, &mut buf);
    buf
}

pub fn fn_item(id: &str, map: &IDMap) -> String {
    if let Some(item) = map.get_item(id) {
        if let ItemEnum::Function(f) = &item.inner {
            let fname = item.name.as_deref().unwrap_or("");
            return parse_fn(&item.visibility, fname, f);
        }
    }
    String::new()
}
