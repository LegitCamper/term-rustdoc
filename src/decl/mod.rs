use crate::tree::IDMap;
use rustdoc_types::{ItemEnum, Visibility};
use std::fmt::Write;

mod fn_;
pub use fn_::fn_item;

mod struct_;

fn vis(v: &Visibility, buf: &mut String) {
    match v {
        Visibility::Public => buf.push_str("pub "),
        Visibility::Default => (),
        Visibility::Crate => buf.push_str("pub(crate) "),
        Visibility::Restricted { path, .. } => _ = write!(buf, "pub({path}) "),
    };
}

/// Parse Item as String.
trait Parse {
    fn parse(&self, v: &Visibility, fname: &str) -> String;
    fn buf(v: &Visibility) -> String {
        let mut buf = String::with_capacity(128);
        vis(v, &mut buf);
        buf
    }
    fn item(item: &ItemEnum) -> Option<&Self>;
    fn item_str(id: &str, map: &IDMap) -> String {
        if let Some(item) = map.get_item(id) {
            if let Some(inner) = Self::item(&item.inner) {
                let fname = item.name.as_deref().unwrap_or("");
                return inner.parse(&item.visibility, fname);
            }
        }
        String::new()
    }
}

pub fn item_str(id: &str, map: &IDMap) -> String {
    if let Some(item) = map.get_item(id) {
        let fname = item.name.as_deref().unwrap_or("");
        let vis = &item.visibility;
        return match &item.inner {
            ItemEnum::Import(reexport) => reexport
                .id
                .as_ref()
                .map(|id| item_str(&id.0, map))
                .unwrap_or_default(),
            ItemEnum::Function(f) => f.parse(vis, fname),
            ItemEnum::Struct(s) => s.parse(vis, fname),
            _ => String::new(),
        };
    }
    String::new()
}
