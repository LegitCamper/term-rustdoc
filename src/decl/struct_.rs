use super::Format;
use crate::{
    tree::IDMap,
    type_name::{generics, short},
};
use rustdoc_types::{Id, ItemEnum, Struct, StructKind, Visibility};
use std::fmt::Write;

fn private(b: bool) -> &'static str {
    const PRIVATE: &str = "/* private fields */";
    if b {
        PRIVATE
    } else {
        ""
    }
}

impl Format for Struct {
    fn parse(&self, v: &Visibility, fname: &str, map: &IDMap) -> String {
        let Struct {
            kind, generics: g, ..
        } = self;

        let mut buf = super::buf(v);
        _ = write!(&mut buf, "struct {fname}");
        let (def, where_) = generics(g);
        let b = &mut buf;
        match (kind, &def, &where_) {
            (StructKind::Unit, None, None) => b.push(';'),
            (StructKind::Unit, None, Some(w)) => _ = write!(b, "where {w};"),
            (StructKind::Unit, Some(d), None) => _ = write!(b, "<{d}>;"), // very unlikely
            (StructKind::Unit, Some(d), Some(w)) => _ = write!(b, "<{d}> where {w};"),
            // TODO: need IDMap to know Path
            (StructKind::Tuple(t), None, None) => _ = write!(b, "({t:?})"),
            (StructKind::Tuple(t), None, Some(w)) => _ = write!(b, "({t:?}) where {w};"),
            (StructKind::Tuple(t), Some(d), None) => _ = write!(b, "<{d}>({t:?});"),
            (StructKind::Tuple(t), Some(d), Some(w)) => _ = write!(b, "<{d}>({t:?}) where {w};"),
            (
                StructKind::Plain {
                    fields,
                    fields_stripped,
                },
                None,
                None,
            ) => write_body(b, fields, map, *fields_stripped),
            (
                StructKind::Plain {
                    fields,
                    fields_stripped,
                },
                None,
                Some(w),
            ) => {
                _ = writeln!(b, " where {w}");
                write_body(b, fields, map, *fields_stripped);
            }
            (
                StructKind::Plain {
                    fields,
                    fields_stripped,
                },
                Some(d),
                None,
            ) => {
                _ = writeln!(b, "<{d}>");
                write_body(b, fields, map, *fields_stripped);
            }
            (
                StructKind::Plain {
                    fields,
                    fields_stripped,
                },
                Some(d),
                Some(w),
            ) => {
                _ = writeln!(b, "<{d}>\nwhere {w}");
                write_body(b, fields, map, *fields_stripped);
            }
        }
        buf
    }
}

fn write_body(b: &mut String, fields: &[Id], map: &IDMap, fields_stripped: bool) {
    b.push_str(" {\n");
    push_fields(fields, map, b);
    let has_public_fields = !fields.is_empty();
    let pr = private(fields_stripped);
    if has_public_fields {
        b.push_str("    ");
    }
    b.push_str(pr);
    if has_public_fields {
        b.push('\n');
    }
    b.push('}');
}

fn field(id: &str, map: &IDMap, buf: &mut String) {
    if let Some(fi) = map.get_item(id) {
        if let Some(name) = fi.name.as_deref() {
            if let ItemEnum::StructField(ty) = &fi.inner {
                _ = writeln!(buf, "    {name}: {},", short(ty));
            }
        }
    }
}

fn push_fields(ids: &[Id], map: &IDMap, buf: &mut String) {
    for id in ids.iter().map(|id| &*id.0) {
        field(id, map, buf);
    }
}
