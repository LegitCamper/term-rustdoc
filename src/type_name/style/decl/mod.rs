mod function;
mod struct_;

use super::{
    path::{FindName, Format, Short},
    StyledType, Vis,
};
use crate::tree::IDMap;
use rustdoc_types::{ItemEnum, Visibility};

pub fn item_styled(id: &str, map: &IDMap) -> StyledType {
    if let Some(item) = map.get_item(id) {
        let vis_name_map = VisNameMap {
            name: item.name.as_deref().unwrap_or(""),
            vis: &item.visibility,
            map,
        };
        let mut buf = StyledType::with_capacity(48);
        match &item.inner {
            ItemEnum::Import(reexport) => {
                let id = reexport.id.as_ref();
                return id.map(|id| item_styled(&id.0, map)).unwrap_or_default();
            }
            ItemEnum::Function(f) => f.format_as_short(vis_name_map, &mut buf),
            ItemEnum::Struct(s) => s.format_as_short(vis_name_map, &mut buf),
            _ => return StyledType::default(),
        };
        return buf;
    }
    StyledType::default()
}

impl Format for Visibility {
    fn format<Kind: FindName>(&self, buf: &mut StyledType) {
        buf.write(match self {
            Visibility::Public => Vis::Pub,
            Visibility::Default => Vis::Default,
            Visibility::Crate => Vis::PubCrate,
            Visibility::Restricted { parent, path } => {
                buf.write_vis_scope(parent, path);
                return;
            }
        });
    }
}

struct VisNameMap<'a> {
    vis: &'a Visibility,
    name: &'a str,
    map: &'a IDMap,
}

trait Declaration {
    fn format<K: FindName>(&self, map: VisNameMap, buf: &mut StyledType);
    fn format_as_short(&self, map: VisNameMap, buf: &mut StyledType) {
        <Self as Declaration>::format::<Short>(self, map, buf);
    }
}
