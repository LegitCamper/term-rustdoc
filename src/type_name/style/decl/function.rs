use super::{Declaration, VisNameMap};
use crate::type_name::style::{
    path::{FindName, Format},
    utils::write_comma,
    Function as Func, Punctuation, StyledType, Syntax,
};
use rustdoc_types::{FnDecl, Function, Generics};

impl Declaration for Function {
    fn format<K: FindName>(&self, map: VisNameMap, buf: &mut StyledType) {
        map.vis.format::<K>(buf);
        let Function {
            decl,
            generics,
            header,
            has_body,
        } = self;
        header.format::<K>(buf);
        buf.write(Func::Fn);
        buf.write(map.name);
        let Generics {
            params,
            where_predicates,
        } = generics;
        if !params.is_empty() {
            params.format::<K>(buf);
        }
        decl.format::<K>(buf);
        where_predicates.format::<K>(buf);
        if !*has_body {
            // if a function has no body, it's likely an associated function in trait definition
            buf.write(Punctuation::SemiColon);
        }
    }
}

impl Format for FnDecl {
    fn format<Kind: FindName>(&self, buf: &mut StyledType) {
        let FnDecl {
            inputs,
            output,
            c_variadic,
        } = self;
        // Multiline for args if the count is more than 2.
        let multiline = (inputs.len() + *c_variadic as usize) > 2;
        buf.write_in_parentheses(|buf| {
            buf.write_slice(
                inputs,
                |arg, buf| {
                    if multiline {
                        buf.write(Punctuation::NewLine);
                        buf.write(Punctuation::Indent);
                    }
                    arg.format::<Kind>(buf);
                },
                write_comma,
            );
            if *c_variadic {
                write_comma(buf);
                if multiline {
                    buf.write(Punctuation::NewLine);
                    buf.write(Punctuation::Indent);
                }
                buf.write(Syntax::Variadic);
            }
            if multiline {
                buf.write(Punctuation::NewLine);
            }
        });
        if let Some(ty) = output {
            buf.write(Syntax::ReturnArrow);
            ty.format::<Kind>(buf);
        }
    }
}
