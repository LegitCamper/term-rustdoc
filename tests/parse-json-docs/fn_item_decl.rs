use crate::{doc, snap};
use term_rustdoc::decl::fn_item;

#[test]
fn fn_items() {
    let map = &doc();
    let dmod = map.dmodule();
    let mut fns_str = Vec::new();
    for fn_ in &dmod.functions {
        fns_str.push(fn_item(&fn_.id, map));
    }
    // for struct_ in &dmod.structs {
    //     for inh in &*struct_.impls.merged_inherent.functions {
    //         fns_str.push(fn_item(&inh.id, map));
    //     }
    // }
    // for enum_ in &dmod.enums {
    //     for inh in &*enum_.impls.merged_inherent.functions {
    //         fns_str.push(fn_item(&inh.id, map));
    //     }
    // }
    // for union_ in &dmod.unions {
    //     for inh in &*union_.impls.merged_inherent.functions {
    //         fns_str.push(fn_item(&inh.id, map));
    //     }
    // }
    // for trait_ in &dmod.traits {
    //     for fn_ in &*trait_.functions {
    //         fns_str.push(fn_item(&fn_.id, map));
    //     }
    // }
    snap!(fns_str, @r###"
    [
        "pub func_dyn_trait(d: &(dyn ATrait + Send + Sync)) -> &dyn ATrait",
        "pub func_dyn_trait2(_: Box<dyn ATrait + Send + Sync>)",
        "pub func_primitive(s: &str) -> usize",
        "pub func_with_1arg(_: FieldsNamedStruct)",
        "pub func_with_1arg_and_ret(f: FieldsNamedStruct) -> AUnitEnum",
        "pub func_with_no_args()",
    ]
    "###);
}

#[test]
fn methods() {
    let map = &doc();
    let dmod = map.dmodule();
    let mut fns_str = Vec::new();
    for struct_ in &dmod.structs {
        for inh in &*struct_.impls.merged_inherent.functions {
            fns_str.push(fn_item(&inh.id, map));
        }
    }
    for enum_ in &dmod.enums {
        for inh in &*enum_.impls.merged_inherent.functions {
            fns_str.push(fn_item(&inh.id, map));
        }
    }
    for union_ in &dmod.unions {
        for inh in &*union_.impls.merged_inherent.functions {
            fns_str.push(fn_item(&inh.id, map));
        }
    }
    for trait_ in &dmod.traits {
        for fn_ in &*trait_.functions {
            fns_str.push(fn_item(&fn_.id, map));
        }
    }
    snap!(fns_str, @r###"
    [
        "pub by_rc(self: Rc<Self>)",
        "pub by_ref(&self)",
        "pub by_ref_mut(&mut self)",
        "pub consume(self)",
        "pub new() -> Self",
    ]
    "###);
}
