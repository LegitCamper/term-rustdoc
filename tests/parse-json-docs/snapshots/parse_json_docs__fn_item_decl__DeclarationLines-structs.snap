---
source: tests/parse-json-docs/fn_item_decl.rs
expression: DisplaySlice(&lines)
---
pub struct ☺️ AUnitStruct#0:3:1787#☺️ ;☺️ 

pub struct ☺️ FieldsNamedStruct#0:17:1800#☺️  {☺️ 
    field1: ☺️ AUnitStruct#0:3:1787#☺️ ,☺️ 
    field2: ☺️ AStructAlias#0:3:1787#☺️ ,☺️ 
    field3: ☺️ Vec#5:7052:263#☺️ <☺️ FieldsNamedStruct#0:17:1800#☺️ >,☺️ 
    /* private fields */☺️ 
}☺️ 

pub struct ☺️ Named#0:113:1859#☺️  {☺️ 
    fut: ☺️ Pin#2:42479:1857#☺️ <☺️ Box#5:294:1815#☺️ <dyn ☺️ Future#2:45136:150#☺️ <Output = ☺️ Result#2:42487:226#☺️ <(), ☺️ Box#5:294:1815#☺️ <dyn ☺️ Error#2:2837:138#☺️ >>>>>,☺️ 
    /* private fields */☺️ 
}☺️ 

pub struct ☺️ NamedAllPrivateFields#0:109:1854#☺️  { /* private fields */ }☺️ 

pub struct ☺️ NamedAllPublicFields#0:111:1856#☺️  {☺️ 
    fut: ☺️ Pin#2:42479:1857#☺️ <☺️ Box#5:294:1815#☺️ <dyn ☺️ Future#2:45136:150#☺️ <Output = ☺️ Result#2:42487:226#☺️ <(), ☺️ Box#5:294:1815#☺️ <dyn ☺️ Error#2:2837:138#☺️ >>>>>☺️ 
}☺️ 

pub struct ☺️ NamedGeneric#0:116:1860#☺️ <'a, T, const N: usize> {☺️ 
    f1: &'a T,☺️ 
    f2: [T; N]☺️ 
}☺️ 

pub struct ☺️ NamedGenericAllPrivate#0:131:1864#☺️ <'a, T, const N: usize> { /* private fields */ }☺️ 

pub struct ☺️ NamedGenericWithBound#0:123:1863#☺️ <'a, T = (), const N: usize = 1>☺️ 
where☺️ 
    T: ☺️ Copy#2:2992:119#☺️ ☺️ 
{☺️ 
    f1: &'a [T],☺️ 
    f2: [T; N]☺️ 
}☺️ 

pub struct ☺️ NamedGenericWithBoundAllPrivate#0:137:1865#☺️ <'a, T, const N: usize>☺️ 
where☺️ 
    T: ☺️ Copy#2:2992:119#☺️ ☺️ 
{ /* private fields */ }☺️ 

pub struct ☺️ Tuple#0:80:1848#☺️ (☺️ 
    _,☺️ 
    _,☺️ 
    ☺️ FieldsNamedStruct#0:17:1800#☺️ ☺️ 
);☺️ 

pub struct ☺️ TupleAllPrivate#0:85:1849#☺️ (_, _, _);☺️ 

pub struct ☺️ TupleGeneric#0:92:1851#☺️ <'a, T: 'a, const N: usize>(☺️ 
    &'a T,☺️ 
    [T; N]☺️ 
);☺️ 

pub struct ☺️ TupleGenericWithBound#0:100:1852#☺️ <'a, T, const N: usize>(☺️ 
    &'a T,☺️ 
    _☺️ 
)☺️ 
where☺️ 
    [T; N]: ,☺️ 
    T: ☺️ Copy#2:2992:119#☺️  + 'a;☺️ 

pub struct ☺️ TupleWithBound#0:90:1850#☺️ ()☺️ 
where☺️ 
    u8: ☺️ Copy#2:2992:119#☺️ ;☺️ 

pub struct ☺️ Unit#0:69:1844#☺️ ;☺️ 

pub struct ☺️ UnitGeneric#0:73:1846#☺️ <const N: bool>;☺️ 

pub struct ☺️ UnitGenericWithBound#0:76:1847#☺️ <const N: usize>☺️ 
where☺️ 
    [(); N]: ;☺️ 

pub struct ☺️ UnitWithBound#0:71:1845#☺️ ☺️ 
where☺️ 
    u8: ☺️ Copy#2:2992:119#☺️ ;☺️ 

pub struct ☺️ AUnitStruct#0:3:1787#☺️ ;☺️ 

pub struct ☺️ AUnitStruct#0:3:1787#☺️ ;☺️
